### PR TITLE
[parallel] fix: stabilize nested FSDP2 backward prefetch topology

### PR DESCRIPTION
**Pull request by Chyu0401**
### What does this PR do?

This PR fixes unbounded FSDP2 parameter all-gather retention observed with the following combination:

```text
activation checkpointing
+ nested FSDP2 groups
+ MoE expert parallelism
+ default backward prefetch
```

PyTorch's default FSDP2 backward prefetch infers the next group from the mutable post-forward execution order. Activation checkpoint recomputation appends nested Decoder and Expert groups to that order again. The heuristic can then prefetch an Expert group that is not consumed by the next `pre_backward`, leaving its all-gather result alive until final backward cleanup. On deep MoE models, one unconsumed Expert buffer may accumulate per layer and cause OOM before backward finishes.

This change assigns every nested FSDP group a stable role at parallelization time and builds independent backward prefetch chains for equivalent roles across sequential transformer blocks. For example:

```text
decoder[i] -> decoder[i - 1]
expert[i]  -> expert[i - 1]
```

The first group in each chain uses a self-target to prevent falling back to the dynamic default heuristic. Backward prefetch configuration is also decoupled from `forward_prefetch`, so disabling forward prefetch no longer disables the safe backward topology.

This PR does not change forward/backward math, parameter sharding, gradient reduction, optimizer updates, or activation checkpoint recomputation.

Related work:

- #871 proposes a broader decoupled-trigger FSDP2 approach for MoE recomputation. This PR keeps the existing composable FSDP2 path and stabilizes its manual prefetch topology.
- #101 reported a previous heterogeneous-layer prefetch failure. The role-based chains in this PR avoid assuming that every layer has the same nested-module layout.

### Checklist Before Starting

- Related PRs/issues:
  - https://github.com/ByteDance-Seed/VeOmni/pull/871
  - https://github.com/ByteDance-Seed/VeOmni/issues/101
- PR title follows the required format:
  - `[parallel] fix: stabilize nested FSDP2 backward prefetch topology`

### Test

Focused unit tests:

```bash
python -m pytest -q tests/distributed/test_fsdp2_prefetch.py
```

Result:

```text
7 passed
```

Code quality checks for the changed files:

```bash
ruff check \
  veomni/distributed/torch_parallelize.py \
  tests/distributed/test_fsdp2_prefetch.py

ruff format --check \
  veomni/distributed/torch_parallelize.py \
  tests/distributed/test_fsdp2_prefetch.py
```

Result:

```text
All checks passed.
2 files already formatted.
```

The same FSDP2 diff was validated in an Ascend NPU integration run:

```text
Model:                  Qwen3.5-35B-A3B
Devices:                4 x Ascend 910B
DP shard size:          4
EP size:                2
Sequence length:        4096
Activation checkpoint: enabled
Forward prefetch:       disabled
FSDP offload:           disabled
Activation offload:     disabled
```

The framework configured three static backward chains covering 107 non-root FSDP groups:

```text
FSDP2 manual prefetch configured for 3 backward chain(s) and 107 group(s);
forward prefetch enabled=False.
```

Pending prefetch memory remained bounded during backward:

```text
Decoder prefetch: approximately 0.0697 GiB
Expert prefetch:  approximately 0.7500 GiB
Total:            approximately 0.8197 GiB
```

Backward completed from the last layer to Layer 0. At `backward_post`:

```text
Decoder groups:           40/40 sharded
Expert groups:            40/40 sharded
Other groups:             27/27 sharded
Root group:                 1/1 sharded
pending_all_gather_groups:    []
```

The run subsequently OOMed during the first `AdamW.step()` while lazily creating `exp_avg_sq`. This is a four-device optimizer-state capacity limitation after backward has completed and is unrelated to FSDP prefetch.

Multi-step 397B validation and a stable-step throughput comparison are still in progress.

### API and Usage Example

No new public API or configuration field is introduced.

The existing option keeps its forward-only meaning:

```yaml
train:
  accelerator:
    fsdp_config:
      forward_prefetch: false
```

For models with nested extra-parallel or mixed-precision-ignored FSDP groups, static backward prefetch chains are still configured when forward prefetch is disabled.

### Design & Code Changes

- Assign each nested FSDP group a stable role using:
  - group kind;
  - extra-parallel name;
  - normalized model-stack path;
  - module-relative path.
- Register overlapping nested modules once by object identity.
- Build one backward chain per role in original model traversal order instead of lexicographic FQN order.
- Configure `group[i] -> group[i - 1]` backward prefetch targets.
- Use a self-target for the first group in each chain to suppress fallback to the mutable default heuristic.
- Configure backward prefetch independently from `forward_prefetch`.
- Preserve the existing forward-prefetch behavior when forward prefetch is enabled.
- Add tests covering:
  - every nested FSDP role;
  - forward/backward prefetch decoupling;
  - singleton self-target chains;
  - heterogeneous blocks;
  - multiple model-stack namespaces;
  - original traversal ordering;
  - duplicate module registration.

The static chain assumes that transformer target modules execute sequentially in model traversal order. Models with dynamic layer skipping, repeated module calls, or non-standard interleaving require separate validation.

### Checklist Before Submitting

- [ ] Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md).
- [ ] Run `pre-commit run --all-files` before final submission.
- [x] Focused Ruff and unit-test checks pass.
- [x] No public API/configuration documentation update is required.
- [x] No `tasks/` training script was moved or renamed.
- [x] Tests are included under the existing `tests/distributed/` collection; no CI workflow change is required.

### CODE REVIEW

**Comment by Lihui-Gu** [comment]
https://github.com/pytorch/pytorch/blob/449b1768410104d3ed79d3bcfe4ba1d65c7f22c0/torch/distributed/fsdp/_fully_shard/_fsdp_param_group.py#L460  torch2.10 solved, no changes to VeOmni are required.
